pytools/graph.py

```diff
@@ -369,4 +370,40 @@ def compute_induced_subgraph(graph: Mapping[T, Set[T]],
 
 # }}}
 
+
+# {{{
+
+def visualize_graph(graph: Mapping[T, Iterable[T]]) -> None:
+    """
+    Visualize the graph *graph*. Needs the ``dot`` program installed.
+
+    :arg graph: A :class:`collections.abc.Mapping` representing a directed
+        graph. The dictionary contains one key representing each node in the
+        graph, and this key maps to a :class:`collections.abc.Set` of nodes
+        that are connected to the node by outgoing edges.
+    """
+    from pytools import UniqueNameGenerator
+    id_gen = UniqueNameGenerator(forced_prefix="mynode")
+
+    from pytools.dot import dot_escape, show_dot
+
+    array_to_id = {}
+
+    res = "digraph mygraph {\n"
+
+    for node, targets in graph.items():
+        array_to_id[node] = id_gen()
+        res += f'{array_to_id[node]} [label="{dot_escape(repr(node))}"];\n'
+        for t in targets:
+            array_to_id[t] = id_gen()
+            res += f'{array_to_id[t]} [label="{dot_escape(repr(t))}"];\n'
+            res += f"{array_to_id[node]} -> {array_to_id[t]};\n"
```

Comment:
This has quadratic complexity because of the inefficient string assembly. Instead, build a list and `join` it.

Reply:
Is 4dd3b81 what you had in mind?

Reply:
Further improved in https://github.com/inducer/pytools/pull/151/commits/48169a565b96dc80a4e5d80453f3aa5e942e27ca